### PR TITLE
cli: Add `--show-origin` option to `config list`

### PR DESCRIPTION
**Pull request by wfchandler**
Add a new `--show-origin` option to `jj config list` to display the
source type and file path (when present) of config values.

Related to https://github.com/martinvonz/jj/issues/1047

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [ ] ~~I have updated the config schema (cli/src/config-schema.json)~~ Not applicable
- [x] I have added tests to cover my changes

### CODE REVIEW

File: CHANGELOG.md
Comment (jonathantanmy) [inline CHANGELOG.md:26]: Thanks, this looks like a useful feature. The equivalent option in Git is called `--show-origin`, and here and in other places, you also describe the information displayed as an "origin", so it might be better to use that term. (You can hold off making that change until other people have chimed in on the name.)

File: cli/src/config.rs
Comment (jonathantanmy) [inline cli/src/config.rs:47]: I would use a custom name for this function (instead of implementing a trait), since we may not always want to display a config source in the same way (for example, we could represent each source type by a single letter). But I'll leave it to people more experienced in the jj project than I to decide.

File: CHANGELOG.md
Comment (wfchandler) [reply CHANGELOG.md:26]: I should have referenced https://github.com/martinvonz/jj/issues/1047 in the PR, will add that now. That has a task to add `--show-sources`, but I'm happy to switch to `--show-origin` if preferred.

File: cli/src/config.rs
Comment (wfchandler) [reply cli/src/config.rs:47]: Hmm, I see the benefit keeping our options open in regards to formatting.  Searching around in `cli/src` I generally see `fmt::Display` being used, e.g. [RemoteBranchName](https://github.com/martinvonz/jj/blob/dce99cf1f83bfc54a4130c4fe2b1c1d0a81a980c/cli/src/cli_util.rs#L2460), though there are exceptions like [CommitOrChangeId](https://github.com/martinvonz/jj/blob/dce99cf1f83bfc54a4130c4fe2b1c1d0a81a980c/cli/src/commit_templater.rs#L564-L580).

Let's see what others say.

**Comment by wfchandler** [comment]
I should mention that on Windows `fs::canonicalize()` returns the [somewhat unusual UNC format](https://github.com/rust-lang/rust/issues/42869):

```
usercfg:\\?\C:\Users\myuser\AppData\Roaming\jj\config.toml       ui.color="auto"
```

The [dunce](https://docs.rs/dunce/latest/dunce/) crate could be used to convert to the more commonly used simple format that users will be more familiar with.

File: cli/tests/test_config_command.rs
Comment (jonathantanmy) [reply cli/tests/test_config_command.rs:173]: Not sure if the colon placement will make it easier to parse - there is already a tab there that parsers can split on.

In Git, looking at the commit that introduced this feature (70bd879), I don't think there was an explicit decision to unify all the files - it was just about exposing what the parsing mechanism already knew. Support for `--show-scope` (to differentiate between the different types of files) was added later.

Comment (martinvonz) [reply cli/tests/test_config_command.rs:173]: > Not sure if the colon placement will make it easier to parse - there is already a tab there that parsers can split on.

True. I think it still helps humans parse it if the colon consistently comes after the source/origin.

File: cli/tests/test_config_command.rs
Comment (yuja) [inline cli/tests/test_config_command.rs:703]: better to normalize the output by `.replace()` or `insta::with_settings!({filters => ..})`. It would be annoying if we had to run tests on both platforms to update the snapshot.

File: cli/src/commands/config.rs
Comment (yuja) [inline cli/src/commands/config.rs:332]: nit: use space instead of tab? I don't think `\t` is used elsewhere.

File: cli/src/config.rs
Comment (yuja) [reply cli/src/config.rs:47]: Other than `Error` types, I think `fmt::Display` is used where there's  one reasonable string representation.

For `ConfigSource` enum, I'm okay with either way. I personally would avoid implementing `fmt::Display` because the output (like `"command line"`) seems more verbose than I would expect from `fmt::Display` of enums, though.

File: cli/tests/test_config_command.rs
Comment (wfchandler) [reply cli/tests/test_config_command.rs]: Good call, I added a filter.

File: cli/tests/test_config_command.rs
Comment (wfchandler) [reply cli/tests/test_config_command.rs]: I've switched to the proposed syntax. To my eye it's a hard to tell where the path ends and the value starts on longer lines.

```
env: ui.pager="/usr/bin/less"
usercfg /home/user/.config/jj/config.toml: ui.editor="nvim"
usercfg /home/user/.config/jj/config.toml: user.email="wfc@wfchandler.org"
usercfg /home/user/.config/jj/config.toml: user.name="Will Chandler"
repocfg /home/user/devel/jj/.jj/repo/config.toml: colors.commit_id prefix.bold=true
repocfg /home/user/devel/jj/.jj/repo/config.toml: revset-aliases.immutable_heads()="main@origin | gh-pages@origin"
cmdline: ui.paginate="never"
```

Also changed `command line` to `cmdline` for easier splitting.

File: cli/src/config.rs
Comment (wfchandler) [reply cli/src/config.rs]: I've added a `description` method in lieu of the `fmt::Display` implementation.

File: CHANGELOG.md
Comment (wfchandler) [reply CHANGELOG.md]: Updated flag name to `--show-origin`.

File: cli/src/commands/config.rs
Comment (wfchandler) [reply cli/src/commands/config.rs]: Done as part of https://github.com/martinvonz/jj/pull/3023#discussion_r1485751209

File: cli/tests/test_config_command.rs
Comment (martinvonz) [reply cli/tests/test_config_command.rs]: > To my eye it's a hard to tell where the path ends and the value starts on longer lines.

Oh, we should use the formatter to label the different parts of the output so it can be styled differently, but that's out of scope for this PR.

File: cli/src/config.rs
Comment (yuja) [inline cli/src/config.rs:599]: > environment config variables are no longer case-sensitive

Oops, it probably means the structured command config no longer works. Environment variables are case sensitive on Unix, and we do use the TOML table syntax to override them.

Maybe it's time to roll our own config module? I think it will help implement #616 and/or handle colors/revset-aliases overrides better.

Sorry for the late reply. I didn't notice this yesterday.

**Review by yuja — CHANGES_REQUESTED** [review]

File: cli/src/config.rs
Comment (wfchandler) [reply cli/src/config.rs:599]: Hmm, I think my summary of what changed in `config` v0.14.0 is not quite right.  Prior to v0.14.0 environment variable keys [were already being downcased](https://github.com/mehcode/config-rs/blob/v0.13.4/src/env.rs#L181), see https://github.com/mehcode/config-rs/issues/340, but the internal keys tracked by `config` were not. For example, an internal `PAGER` key would not be set by environment variable `PAGER` because the latter had been converted to `pager`.

The change in https://github.com/mehcode/config-rs/pull/354 was to downcase the internal keys as well. Looking at the config schema I don't see any mixed case settings, do you have an example of what you would expect to break so I can be sure I understand?

File: cli/tests/test_config_command.rs
Comment (wfchandler) [reply cli/tests/test_config_command.rs]: Makes sense 👍

File: cli/src/config.rs
Comment (yuja) [reply cli/src/config.rs:599]: > Prior to v0.14.0 environment variable keys [were already being downcased](https://github.com/mehcode/config-rs/blob/v0.13.4/src/env.rs#L181)

I think you're talking about the config loaded from environment variables, but this `CommandNameAndArgs` stuff is data configured in a config file, and passed to the external process. So the case matters.

> do you have an example of what you would expect to break

In addition to this, `[aliases]`, `[revset-aliases]`, `[template-aliases]`, `[merge-tools.<NAME>]` would break if keys are downcased. They are free-form dictionary, and users may define `revset-aliases.HEAD = "@-"` for example.

Comment (martinvonz) [reply cli/src/config.rs:599]: @wfchandler: To clarify, I think Yuya is talking about https://github.com/martinvonz/jj/blob/26528091e646a79a3724997f06c4af69eaf70297/cli/src/config/misc.toml#L9 (the user can have their own similar config). (Correct me if I'm wrong, @yuja.)

Comment (yuja) [reply cli/src/config.rs:599]: Perhaps, this is the upstream issue. It doesn't describe the exact problem we have, but the root cause would be the same.
https://github.com/mehcode/config-rs/issues/531

Comment (wfchandler) [reply cli/src/config.rs:599]: Thanks for the explanation, I see what you mean.  I'll go ahead and close out this PR given the problem with `config` v0.14.0.

Thank you @yuja @martinvonz and @jonathantanmy for the thorough review, much better to have caught this issue now.

Comment (yuja) [reply cli/src/config.rs:599]: I hope the problem will be addressed at config-rs, and this PR can be revisited later. Thanks anyways.

Comment (martinvonz) [reply cli/src/config.rs:599]: Yes, thanks for trying, and sorry about the trouble. And thanks to Yuya for noticing the problem!

**Pull request closed without merging**
